### PR TITLE
[Relay][VM][Interpreter] Enable first-class constructors in VM and interpreter via eta expansion

### PR DESCRIPTION
**Pull request by weberlo**
Previously, the VM and interpreter did not support constructors as first-class objects.  For example, the following code snippet would not compile:
```
def @main[A]() -> (fn(A, List[A]) -> List[A]) {
  Cons
}
```
.  The eta expansion pass in this PR remedies this problem by wrapping the constructor in a function, like so:
```
def @main[A]() -> (fn(A, List[A]) -> List[A]) {
  fn (%x: A, %xs: List[A]) -> List[A] {
    Cons(%x, %xs)  
  }
}
```
.  There was an existing eta expansion pass, but it was not recursive and only operated on global variables.  The new pass is configurable, so you can choose to expand constructors, global variables, or both.

Also included are the following minor changes:

- Fix `LambdaLift` pass so it closes over free type variables.
- Fix `LambdaLift` pass so it reuses lifted functions that are alpha equal.
- Fix pretty printing of constructors to be context-sensitive (e.g., print `Cons(A, List[A])` in a type definition but only print `Cons` when in an expr).
- Rename `Module::HasDef` to `Module::ContainGlobalTypeVar` to match the analogous `Module::ContainGlobalVar` method.
- Change prelude definitions to use first-class constructors where applicable.

CC @jroesch @MarisaKirisame @icemelon9

### CODE REVIEW

**Comment by weberlo** [comment]
@jroesch @icemelon9 CI is green.  Take a look when you get a chance.

File: src/relay/pass/eta_expand.cc [hunk 1]
```diff
@@ -20,57 +20,147 @@
...
+ * \brief mutator to perform eta expansion on all functions in a module
+ */
+class EtaExpander : public ExprMutator {
+ public:
+  explicit EtaExpander(
+    const Module& mod,
+    bool expand_constructor,
+    bool expand_global_var)
+      : mod_(mod)
+      , type_var_replacer_(TypeVarReplacer())
```
Comment (icemelon9) [inline src/relay/pass/eta_expand.cc]: move the comma to previous line?

**Comment by weberlo** [comment]
It looks like the tests are timing out.  I think [this addition](https://github.com/apache/incubator-tvm/pull/4218/files#diff-a90b7229a4e9fb69bb84bb8f75c162e8R793) to the interpreter is the cause.  Might need to roll back that change and the change to the prelude, but keep the `EtaExpand` functionality for the VM.  If we want first-class constructors in the interpreter, we'll need to directly support it there.

**Comment by weberlo** [comment]
nvm. there seem to be CI timeouts repo-wide (https://discuss.tvm.ai/t/remove-unused-function-in-relay-module/4753)

**Comment by weberlo** [comment]
@icemelon9 CI is green again

**Comment by icemelon9** [comment]
Thanks @weberlo @MarisaKirisame. This is now merged.

**Pull request merged**